lang/cs/Org.Apache.REEF.Wake/Remote/Impl/WritableLink.cs

```diff
@@ -0,0 +1,279 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Org.Apache.REEF.Tang.Exceptions;
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Util;
+
+namespace Org.Apache.REEF.Wake.Remote.Impl
+{
+    /// <summary>
+    /// Represents an open connection between remote hosts. This class is not thread safe
+    /// </summary>
+    /// <typeparam name="T">Generic Type of message. It is constrained to have implemented IWritable and IType interface</typeparam>
+    public class WritableLink<T> : ILink<T> where T : IWritable
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof (WritableLink<T>));
+
+        private readonly IPEndPoint _localEndpoint;
+        private bool _disposed;
+        private readonly NetworkStream _stream;
+        
+        /// <summary>
+        /// Cache structure to store the constructor functions for various types.
+        /// </summary>
+        private readonly TypeCache<T> _cache; 
+
+        /// <summary>
+        /// Constructs a Link object.
+        /// Connects to the specified remote endpoint.
+        /// </summary>
+        /// <param name="remoteEndpoint">The remote endpoint to connect to</param>
+        public WritableLink(IPEndPoint remoteEndpoint)
+        {
+            if (remoteEndpoint == null)
+            {
+                throw new ArgumentNullException("remoteEndpoint");
+            }
+
+            Client = new TcpClient();
+            Client.Connect(remoteEndpoint);
+
+            _stream = Client.GetStream();
+            _localEndpoint = GetLocalEndpoint();
+            _disposed = false;
+            _cache = new TypeCache<T>();
+        }
+
+        /// <summary>
+        /// Constructs a Link object.
+        /// Uses the already connected TcpClient.
+        /// </summary>
+        /// <param name="client">The already connected client</param>
+        public WritableLink(TcpClient client)
+        {
+            if (client == null)
+            {
+                throw new ArgumentNullException("client");
+            }
+
+            Client = client;
+            _stream = Client.GetStream();
+            _localEndpoint = GetLocalEndpoint();
+            _disposed = false;
+            _cache = new TypeCache<T>();
+        }
+
+        /// <summary>
+        /// Returns the local socket address
+        /// </summary>
+        public IPEndPoint LocalEndpoint
```

Comment:
OK. Whenever we expose implementation details via an API, we make it harder to evolve the code. We should definitely look into this again.